src/python/pants/scm/subsystems/changed.py

```diff
@@ -58,15 +59,43 @@ def register_options(cls, register):
              help='Stop searching for owners once a source is mapped to at least one owning target.')
 
 
-class ScmWrapper(datatype([('scm', SubclassesOf(Scm))])): pass
+class ScmResult(datatype([('scm', SubclassesOf(Scm))])): pass
+
+
+class ScmRequest(datatype([('reason_for_scm', text_type)])): pass
+
+
+class ScmRequestFailed(Exception): pass
+
+
+@rule(ScmResult, [Select(ScmRequest)])
+def try_get_scm(scm_request):
```

Comment:
I think that leaving the wrapped SCM as a parameter is a bit cleaner, for the reason mentioned on #7350.